src/index/posting_list.rs

```diff
@@ -31,17 +31,20 @@ impl PostingList {
             kth_score,
         }
     }
-    const fn map_kth_value(value: usize) -> usize {
+    const fn map_kth_value(value: usize) -> Option<usize> {
         match value {
-            10 => 0,
-            100 => 1,
-            1000 => 2,
-            _ => value, // default case
+            10 => Some(0),
+            100 => Some(1),
+            1000 => Some(2),
+            _ => None, // default case
         }
     }
 
     pub fn kth(&self, k: usize) -> u8 {
-        return self.kth_score[Self::map_kth_value(k)];
+        if let Some(v) = Self::map_kth_value(k) {
```

Comment:
Could be:
```rust
Self::map_kth_value(k).map(|idx| self.kth_score[idx]).unwrap_or(0)
```

But to @JMMackenzie's point, I think he means something like this:

```rust
pub fn kth(&self, k: usize) -> u8 {
    if k <= 1000 {
        self.kth_score[(k - 1).ilog10()]
    } else {
        0
    }
}
```

Notice I used log to make it succinct, but I would encourage you to benchmark if this makes sense.

Either way, even if you are using `match`/`if`, I wouldn't use `map_kth_value`, just straight up use a match statement and put `self.kth_score[...]` in each branch. The reason is that you'll need to do another check for `Option`, and _then_ range check for array access. If anything, I would use `self.kth_score.get(map_kth_value(k)).unwrap_or(0)` and let the result of the inner func go out of range. This way, you should be able to only have one check, I think.

In fact, the function I suggested above could be:

```rust
pub fn kth(&self, k: usize) -> u8 {
    self.kth_score.get((k - 1).ilog10()).unwrap_or(0)
}
```

Though now you must calculate log first, even for large `k`. But if this is not an issue, you can hardly beat the simplicity of a one-liner ;)

Reply:
Or how about this crazy idea:

```rust
pub fn kth(&self, k: usize) -> u8 {
    let idx = ((n > 10) as usize)
        + (n > 100) as usize
        + (n > 1000) as usize;
    self.kth_score.get(idx).unwrap_or(0)
}
```

Reply:
That's beautiful 😍